graph/generate.go

```diff
@@ -0,0 +1,254 @@
+/*
+Package graph provides core lineage graph services. It defines the graph datatype and provides the "New"
+function for creating one.
+*/
+package graph
+
+import (
+	"fmt"
+
+	mbp "github.com/GoogleCloudPlatform/healthcare-data-harmonization/mapping_engine/proto"
+	proto "github.com/golang/protobuf/proto"
+)
+
+var idFactory IDfactory = &autoIncFactory{currentID: 0}
+
+// setIDfactory sets the interface used for ID generation
+func setIDfactory(factory IDfactory) {
+	idFactory = factory
+}
+
+// msgContext is a wrapper around a proto message storing additional context, like parent node.
+type msgContext struct {
+	msg        proto.Message
+	descendant Node
+}
+
+type stack []msgContext
+
+func (s stack) push(mc msgContext) stack {
+	return append(s, mc)
+}
+
+// reversePush pushes an array of proto messages onto the stack in reverse order. It assigns them all the same context.
+// The order is reversed to preserve the ordering of locally-defined whistler targets. For example, an array
+// of FieldMappings extracted from a Projector
+func (s stack) reversePush(msgs []proto.Message, descendant Node) stack {
+	for i := len(msgs) - 1; i >= 0; i-- {
+		s = s.push(msgContext{
+			msg:        msgs[i],
+			descendant: descendant,
+		})
+	}
+	return s
+}
+
+func (s stack) pop() (msgContext, stack) {
+	n := len(s) - 1
```

Comment:
What will happen if someone tries to pop from an empty stack? Should we guard against it?

Reply:
pop() now returns (msgContext, bool) to indicate whether popping was successful. It fails if the stack is empty.